#527 - Status fixes

shravanshetty1 opened the pull request.

shivdudhani commented on cmd/kyverno/main.go:
```diff
@@ -135,12 +137,19 @@ func main() {
+	// Policy Status Handler - deals with all logic related to policy status
+	statusSync := policyStatus.NewSync(
+		pclient,
+		stopCh,
```
Pass the `stopCh` to the `Run` method instead of storing it in Sync

shivdudhani commented:
How are we capturing the processing Times for processing generate Rules?
In `pkg/generate/generate.go` we process the rules and create the resources in `applyGeneratePolicy` and we need to capture the above times.

shravanshetty1 commented on cmd/kyverno/main.go:
```diff
@@ -135,12 +137,19 @@ func main() {
+	// Policy Status Handler - deals with all logic related to policy status
+	statusSync := policyStatus.NewSync(
+		pclient,
+		stopCh,
```
Will do, but how does this make a difference?

shravanshetty1 commented on pkg/policyStatus/0_main.go:
```diff
@@ -0,0 +1,87 @@
+
+	wait.Until(s.updatePolicyStatus, 2*time.Second, s.stop)
+	<-s.stop
+	s.updatePolicyStatus()
```
It is cleaning up the status cache between to updates, and not accepting new status update request. Since we have implemented graceful shutdown, we should enable these kind of features - otherwise there is no difference between graceful shutdown and force killing the task.

shivdudhani replied:
From implementation point of view, we only react to signals(SIGTERM & SIGINT) for stopCh. With `updatePolicyStatus` we are updating the status from then channel. My concern was this is a best-effort scenario, so if the interrupt is received and we have time to cleanUp then we will capture the stats. But if we receive an interrupt and don't have time to cleanUp the end result will be different, this is more like a post-shutdown step. But in the interest of time, we can let this be and reevaluate if needed.

shravanshetty1 replied:
requested changes have been made

shravanshetty1 commented on pkg/generate/status.go:
```diff
@@ -34,11 +36,17 @@ func (sc StatusControl) Failed(gr kyverno.GenerateRequest, message string, genRe
 	gr.Status.GeneratedResources = genResources
 
+	if grCopy.Status.State != kyverno.Completed {
+		go sc.policyStatus.UpdatePolicyStatusWithGeneratedResourceCount(gr)
```
By using a goroutine, there will be no blocking as the application scales up. However if we use buffered channels, we need to agree on the size of of the channel - and as the application scales, there will be blocking at some point. Go routine is not like a thread and takes insignificant amount of resource, every webhook request is in its own go routine...

Can you provide data suggesting the performance difference between creating a go routine and creating a buffer channel of n length so we can clearly see why one is better over the other?

shivdudhani replied:
The concern is with creating goroutines with no control over the number of goroutines that will be created. 
- So we don't have control over the number of goroutines
- We are creating goroutines and no waiting for them to complete, there is no waitGroup of anything to verify if the goroutine was complete or failed. It gets difficult to track and debug.
- It does allow us to scale, but its uncontrolled-scalability.

Regarding the data, I do not have performance difference between creating go routines vs buffers as it's internal compute and depends on many factors. It was a suggestion with the above points in mind and design perspective.


Again in the interest of time, instead of delaying the PR if we have control over goroutines been created or if there are plans to add such control in the future I'm fine with the logic.

shravanshetty1 replied:
requested changes have been made

shravanshetty1 commented on pkg/generate/status.go:
```diff
@@ -34,11 +36,17 @@ func (sc StatusControl) Failed(gr kyverno.GenerateRequest, message string, genRe
 
 // Success sets the gr status.state to completed and clears message
 func (sc StatusControl) Success(gr kyverno.GenerateRequest, genResources []kyverno.ResourceSpec) error {
+	grCopy := gr.DeepCopy()
```
Will change this

shravanshetty1 commented on pkg/policyStatus/0_main.go:
```diff
@@ -0,0 +1,87 @@
+
+type Sync struct {
+	cache       *cache
+	listener    chan statusUpdater
```
Will update current design - so that user can send the data and define the logic that will be used on the data

shravanshetty1 replied:
Will update current design - so that user can send the data and define the logic that will be used on the data

shravanshetty1 replied:
@shivdudhani I have decouple sender and reciever as you have asked, please confirm if this is what you meant

shivdudhani replied:
The channel is been used to send the interface object(pointer to implementation and object) and not the actual stats object. 
The sender has to define the logic for aggregation of stats here in webhook pkg, while it should be just sending stats and the status sync should be aggregating it, the webhook need implementation has all access to cache and locks). 

The Sync is just running the methods for the caller, the suggestion was to just send the data and not a pointer to data or interface. 
So the sender only has access to an interface that allows adding to the channel.
The receiver will then read from the channel and define the processing for each type.

The channel should be used to send the data, not a function object which the sync hander just runs.

shravanshetty1 replied:
further redesigned implementation - so that package variables are not used accross packages

shravanshetty1 commented:
@shivdudhani
engine.Generate should create the resource definition and add it the engine response. The sync should take the definition and actually create it.
Once we move the logic of creating the resource definition from applyRule and into engine.generate - we can more accurately calculate rule processing time.

shravanshetty1 commented on pkg/policyStatus/generateStats.go:
```diff
@@ -0,0 +1,85 @@
+}
+
+func (s *Sync) UpdateStatusWithGenerateStats(resp response.EngineResponse) {
+	s.listener <- &generateStats{
```
will change this

shivdudhani commented:
> @shivdudhani
> engine.Generate should create the resource definition and add it the engine response. The sync should take the definition and actually create it.
> Once we move the logic of creating the resource definition from applyRule and into engine.generate - we can more accurately calculate rule processing time.
> 
> Do you agree with this solution?

As per the current design `engine.Generate` does not create the resources but only verify if the resource can be applied to the policy, this was done to delay the resource creation based on the policy after verifying the trigger resource has been generated.

The above suggestion was the previous flow which we modified to current to decouple components.

Also with this PR lets focus on getting the results from `generate` rule process while creating resources(this is yet not captured, the time required to create generated resources ) and handle the refactoring of code and making improvements in a different PR.
It seems we are diverging from the goal of the issue( we already had stats for mutate & validation, only generate was missing and need to be added) but now we have reworked all components which are not needed in this issue.

shravanshetty1 commented:
> > @shivdudhani
> > engine.Generate should create the resource definition and add it the engine response. The sync should take the definition and actually create it.
> > Once we move the logic of creating the resource definition from applyRule and into engine.generate - we can more accurately calculate rule processing time.
> > Do you agree with this solution?
> 
> As per the current design `engine.Generate` does not create the resources but only verify if the resource can be applied to the policy, this was done to delay the resource creation based on the policy after verifying the trigger resource has been generated.
> 
> The above suggestion was the previous flow which we modified to current to decouple components.
> 
> Also with this PR lets focus on getting the results from `generate` rule process while creating resources(this is yet not captured, the time required to create generated resources ) and handle the refactoring of code and making improvements in a different PR.
> It seems we are diverging from the goal of the issue( we already had stats for mutate & validation, only generate was missing and need to be added) but now we have reworked all components which are not needed in this issue.

Current implementation records processing done during `applyGeneratePolicy`
requested changes have been made

shivdudhani commented:
update the package name from `policyStatus` to `policystatus` (consistent with other packages)

shivdudhani commented:
Expose an interface that allows adding values to the channel, instead of exposing the channel itself.
As we sharing pointer to sync, we are expossing other methods like Run, which is not needed.

shivdudhani commented:
Overview:
- Every caller will define the function to parse the status and update the result
- Pass the function pointer via an interface to the channel
- receive on the channel, will lock the map and pass the current record to the above function object. The function object will transform the resource and return the result to be updated in the map. Release the lock.

Future Changes:
- Add more comments to the implementation, as the logic is pretty complicated to understand and define conditions for new users to update the stats.
- Add more debug log statements. (question, the logger used by the function will be the instance of which package? (as it's defined by someone else and run by someone else) )
- Locking the stats for each policy record. Now we lock the whole map, so stats for different policies cannot be done together.

shravanshetty1 commented:
> Expose an interface that allows adding values to the channel, instead of exposing the channel itself.
> As we sharing pointer to sync, we are expossing other methods like Run, which is not needed.

will change but this is being done everywhere, including grgen, resourceWebhookWatcher and others....

shravanshetty1 commented:
> Future Changes:
> 
> * Add more comments to the implementation, as the logic is pretty complicated to understand and define conditions for new users to update the stats.
> * Add more debug log statements. (question, the logger used by the function will be the instance of which package? (as it's defined by someone else and run by someone else) )
> * Locking the stats for each policy record. Now we lock the whole map, so stats for different policies cannot be done together.

1. I have added comments
2. Added basic log tracking every status update - further logs can be added by the caller, i did not see it was needed, hence further logs have not been added
3. Have refactored the locks such that it allows status to be updated for different policies at the same time while ensuring the status for same policies are updated one at a time. I have added a testcase, such that you can easily test this behaviour. Feel free to remove the locks and re run the testcase to confirm any doubts.

Requested changes have been made

shravanshetty1 commented:
> Expose an interface that allows adding values to the channel, instead of exposing the channel itself.
> As we sharing pointer to sync, we are expossing other methods like Run, which is not needed.

requested change has been made

shravanshetty1 commented:
> update the package name from `policyStatus` to `policystatus` (consistent with other packages)

requested change has been made
